### PR TITLE
Add rollback function support

### LINKED ISSUE #3488: Add rollback function support
# This is a Feature Proposal

## Description

https://github.com/serverless/serverless/pull/2495 introduced the rollback support for the whole service stack as described in https://github.com/serverless/serverless/issues/2412.

It would be really helpful to have a rollback functionality on a per function level so that users can rollback a single function rather than the whole stack.

https://github.com/serverless/serverless/pull/2672 implemented such a functionality.

### 1. Use SDK call to update Lambda versioning and alias

We could utilize the Lambda versioning and alias to update the functions version (as it's done in #2672).

**Upsides:**

- Most "natural" solution

**Downsides:**

- Users can be opted-out of Lambda versioning
- Stacks state gets out of sync

---

### 2. Doing `deploy function` under the hood

Overwrite existing code by directly pushing new function code.

**Upsides:**

- Easy implementation
- Multi-provider ready out of the box (?!)
- Fast

**Downsides:**

- Stacks state gets out of sync
- We're not utilizing provider specific features (such as Lambda versioning)

---

### 3. Using `deploy` under the hood

Overwrite existing code by updating the whole stack.

**Upsides:**

- Easy to implement since most of the functionality is already there
- Stack state is in sync

**Downsides:**

- Slow
- Might overwrite other resources the user has changed in his `serverless.yml` file --> Unsafe

---

### 4. Doing a "partial" `deploy`

Modify current CloudFormation template in `.serverless` directory with the new version and code and deploy it behind the scenes.

**Upsides:**

- Going all the way through CloudFormation

**Downsides:**

- Slow
- Stack might be out of sync
- More like a hack

---

### X. Using `rollback` / `deploy` in combination with CloudFormation changesets

Using AWS CloudFormation changesets to let users rollback to a specific function and plan / see corresponding stack changes.

More like a generic `deploy` / `rollback` enhancement rather than a `rollback function` specific proposal.

**Upsides:**

- User can see / plan the changes before rolling back the function
- Stack will be in a synced state after rolling back

**Downsides:**

- Slow
- We have to re-write the deployment to fully support changesets
  - Will be a breaking change

---

/cc @eahefnawy @brianneisler

### PR DESCRIPTION
**Pull request by pmuens**
## What did you implement:

Closes #3488

Add support for rolling back a single function.

Standing on the shoulders of giants here --> Shoutouts to @marcy-terui who did the heavy lifting in ( #2672) 🙌 

## How did you implement it:

Add a new `rollback function` command with its corresponding plugin.

## How can we verify it:

1. Deploy a service
2. Run `serverless deploy list functions` to see all the functions and their versions
3. Run `serverless rollback function -f <name-of-the-function> -v <version>` to rollback a function to the requestd version

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Update the messages below
- [x] Re-run shrinkwrap (because `node-fetch` was bumped)

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

File: lib/plugins/aws/deployList/index.js [hunk 1]
```diff
@@ -61,6 +66,77 @@ class AwsDeployList {
...
+      });
+  }
+
+  getFunctionVersions(funcs) {
+    const requestPromises = [];
+
+    funcs.forEach((func) => {
+      const params = {
+        FunctionName: func.FunctionName,
+        MaxItems: 5,
```
Comment (eahefnawy) [inline lib/plugins/aws/deployList/index.js]: why only 5?

File: lib/plugins/aws/rollbackFunction/index.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        .then(this.restoreFunction),
+    };
+  }
+
+  getFunctionToBeRestored() {
+    const funcName = this.options.function;
+    let funcVersion = this.options.version;
+
+    // versions need to be string so that AWS understands it
+    funcVersion = String(this.options.version);
```
Comment (eahefnawy) [inline lib/plugins/aws/rollbackFunction/index.js]: We should add validation to this option like we do with stage/region and other options. I can run something like `sls rollback function -f hello -v whatever

File: lib/plugins/rollback/index.js [hunk 3]
```diff
@@ -22,6 +22,29 @@ class Rollback {
...
+                required: true,
+              },
+              stage: {
+                usage: 'Stage of the function',
+                shortcut: 's',
+              },
+              region: {
+                usage: 'Region of the function',
+                shortcut: 'r',
+              },
```
Comment (eahefnawy) [inline lib/plugins/rollback/index.js]: I think you're missing the `--version` option here

File: lib/plugins/aws/deployList/index.js (hunk 1 above)
Comment (pmuens) [inline lib/plugins/aws/deployList/index.js]: We only keep 5 versions of the artifacts in the S3 bucket so this aligns with that.

We could display more though...

File: lib/plugins/aws/rollbackFunction/index.js (hunk 2 above)
Comment (pmuens) [inline lib/plugins/aws/rollbackFunction/index.js]: Good point. I've added a validation for that below https://github.com/serverless/serverless/blob/512862fce00a80583cd92d73db8d0ea0630acdfc/lib/plugins/aws/rollbackFunction/index.js#L80-L91

The problem is that you could also provide `$LATEST` so it's not always a number.

File: lib/plugins/rollback/index.js (hunk 3 above)
Comment (pmuens) [inline lib/plugins/rollback/index.js]: Nice catch! 👍

**Comment by pmuens** [comment]
@eahefnawy thanks for the review!

I answered the questions and updated the code according to your feedback.

Only missing piece is the validation of the version. Not sure if we can / should do this since it could be `$LATEST`, `1` and AWS treats them as strings. Not 100% sure if the user can configure the versioning schema so that other values (not only numbers) can be used. 🤔

File: lib/plugins/rollback/index.js (hunk 3 above)
Comment (pmuens) [inline lib/plugins/rollback/index.js]: Done!

**Comment by HyperBrain** [comment]
@pmuens The versions are only set by AWS in a strictly increasing fashion. You cannot change the versioning format. A possible solution for consistent rollbacks would be to store the actually deployed and associated version of a specific deployment into S3 after deployment has been finished. Then you at least know what version represents the S3 folder version.

Imo function based rollback additionally bears an additional problem: How do you guarantee that the function configuration that is deployed, still fits into the CF stack in regards to user resources, etc.?

**Comment by pmuens** [comment]
@HyperBrain thanks for the explanation about the versioning schema!

> Imo function based rollback additionally bears an additional problem: How do you guarantee that the function configuration that is deployed, still fits into the CF stack in regards to user resources, etc.?

Yes, that's also something. we've discussed. The conclusion was that this rollback "only" touches the code is just temporary since it will be replaced with a full deployment once someone runs `serverless deploy`. But yes, that's a valid problem.

**Pull request merged**